src/NumericalAlgorithms/LinearSolver/Convergence.cpp

```diff
@@ -0,0 +1,149 @@
+
+#include "ErrorHandling/Assert.hpp"
+#include "ErrorHandling/Error.hpp"
+#include "NumericalAlgorithms/LinearSolver/IterationId.hpp"
+
+namespace LinearSolver {
+
+ConvergenceCriteria::ConvergenceCriteria(
+    const double max_iterations_in, const double absolute_residual_in,
+    const double relative_residual_in) noexcept
+    : max_iterations(max_iterations_in),
+      absolute_residual(absolute_residual_in),
+      relative_residual(relative_residual_in) {}
+
+void ConvergenceCriteria::pup(PUP::er& p) noexcept {
+  p | max_iterations;
+  p | absolute_residual;
+  p | relative_residual;
+}
+
+bool operator==(const ConvergenceCriteria& lhs,
+                const ConvergenceCriteria& rhs) noexcept {
+  return lhs.max_iterations == rhs.max_iterations and
+         lhs.absolute_residual == rhs.absolute_residual and
+         lhs.relative_residual == rhs.relative_residual;
+}
+
+bool operator!=(const ConvergenceCriteria& lhs,
+                const ConvergenceCriteria& rhs) noexcept {
+  return not(lhs == rhs);
+}
+
+std::ostream& operator<<(std::ostream& os,
+                         const ConvergenceReason& convergence_reason) noexcept {
+  switch (convergence_reason) {
+    case ConvergenceReason::MaxIterations:
+      return os << "MaxIterations";
+    case ConvergenceReason::AbsoluteResidual:
+      return os << "AbsoluteResidual";
+    case ConvergenceReason::RelativeResidual:
+      return os << "RelativeResidual";
+    default:
+      ERROR("Unknown convergence reason");
+  }
+}
+
+boost::optional<ConvergenceReason> convergence_criteria_match(
+    const ConvergenceCriteria& convergence_criteria,
+    const IterationId& iteration_id, const double residual_magnitude,
+    const double initial_residual_magnitude) noexcept {
+  if (residual_magnitude <= convergence_criteria.absolute_residual) {
+    return ConvergenceReason::AbsoluteResidual;
+  }
+  if (residual_magnitude / initial_residual_magnitude <=
+      convergence_criteria.relative_residual) {
+    return ConvergenceReason::RelativeResidual;
+  }
+  if (iteration_id.step_number >= convergence_criteria.max_iterations) {
+    return ConvergenceReason::MaxIterations;
+  }
+  return boost::none;
+}
+
+HasConverged::HasConverged(const ConvergenceCriteria& convergence_criteria,
+                           const IterationId& iteration_id,
+                           const double residual_magnitude,
+                           const double initial_residual_magnitude) noexcept
+    : reason_(convergence_criteria_match(convergence_criteria, iteration_id,
+                                         residual_magnitude,
+                                         initial_residual_magnitude)),
+      convergence_criteria_(convergence_criteria),
+      iteration_id_(iteration_id),
+      residual_magnitude_(residual_magnitude),
+      initial_residual_magnitude_(initial_residual_magnitude) {}
+
+ConvergenceReason HasConverged::reason() const noexcept {
+  ASSERT(reason_,
```

Comment:
I think this should be an `ERROR` because the `*reason_` will throw and I've noticed it's usually quite difficult to figure out what happened when exception is thrown but not caught. I also don't expect that we will pay much in CPU time for changing this to an ERROR and I can see it making somebody's life easier :)

Reply:
I'm not sure if I understand correctly. The `ASSERT` tests if the `boost::optional<ConvergenceReason> reason_` evaluates to `true`, in which case we are sure that unpacking it with `*` will succeed (and never throw).

Reply:
okay